File: services/market_pulse.py
"""Market Pulse — the read model behind the live crypto command center.

This module owns no networking and no cache of its own. It composes the two
market foundations that already exist:

  * ``services.market_data``                  — the 50-row board (rank, logo,
    price, 24h change, market cap, volume, sparkline) plus real price history.
  * ``services.pulse_briefings.crypto_provider`` — the only place in the product
    that calls ``/global`` and ``/search/trending``, i.e. the only honest source
    for the global strip and for *actual* trending.

Both are already polled for the dashboard board and for Pulse Briefings, and
both already stretch their TTLs under ``coingecko_client.effective_ttl``. So
Market Pulse adds zero new provider polling and zero new budget pressure: N
users still cost one call per window, the same call the rest of the product was
making anyway. A fourth cache here would have been a fourth thing to teach about
the credit guard.

Everything leaves here as a PulseSoc market object — ``{id, symbol, name, rank,
price, change24h, marketCap, volume24h, image, updatedAt}`` — so no CoinGecko
field name reaches a client. Absence is null, never zero: a price of 0.00 and a
0.00% change both read as facts on a screen.
"""

from __future__ import annotations

import logging
import time
from typing import Any

from services import market_data
from services import market_intelligence
from services.pulse_briefings import crypto_provider

LOGGER = logging.getLogger(__name__)

# Beyond this the strip stops claiming to be live. It matches the provider's own
# hard serve limit so the two cannot disagree about what "stale" means.
STALE_AFTER_SECONDS = crypto_provider.STALE_MAX_SECONDS

CATEGORIES = ("all", "gainers", "losers", "trending", "watchlist")

# Chip -> board sort. Only categories the canonical provider data can actually
# answer. "trending" is deliberately absent: it is real /search/trending data,
# not a sort of the board, and labelling a big mover "trending" would be a
# claim the provider never made.
_CATEGORY_SORT = {
    "all": "top_market_cap",
    "gainers": "gainers",
    "losers": "losers",
    "watchlist": "top_market_cap",
}


def _num(value: Any) -> float | None:
    if value is None:
        return None
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return number


def _int(value: Any) -> int | None:
    if value is None:
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def normalize_asset(row: dict[str, Any] | None, updated_at: str | None = None) -> dict[str, Any] | None:
    """One board row as a PulseSoc market object, or None if it has no identity."""
    if not isinstance(row, dict):
        return None
    symbol = str(row.get("symbol") or "").upper()
    coin_id = str(row.get("id") or "").strip()
    if not symbol and not coin_id:
        return None
    return {
        "id": coin_id or symbol.lower(),
        "symbol": symbol,
        "name": str(row.get("name") or symbol),
        "rank": _int(row.get("market_cap_rank")),
        "price": _num(row.get("price")),
        "change24h": _num(row.get("change_24h")),
        "marketCap": _num(row.get("market_cap")),
        "volume24h": _num(row.get("volume_24h")),
        "image": str(row.get("image") or ""),
        # Free with the board request — a 50-row list still costs one call.
        "sparkline": [p for p in (row.get("sparkline") or []) if isinstance(p, (int, float))],
        "updatedAt": updated_at,
    }


def _board(category: str = "all", limit: int = 50) -> dict[str, Any]:
    sort_key = _CATEGORY_SORT.get(category, "top_market_cap")
    try:
        # One shared board. Sorting happens inside market_data against the
        # cached rows, so changing chips never costs a provider call.
        return market_data.live_market_board(category=sort_key, limit=max(1, min(int(limit or 50), 80)))
    except Exception as exc:  # noqa: BLE001 - the board owns its own fallback ladder
        LOGGER.info("Market Pulse board unavailable: %s", exc)
        return {"source": "unavailable", "markets": [], "warning": "Market data is temporarily unavailable.",
                "age_seconds": None, "observed_epoch": None, "updated_at": None, "summary": {}}


def _freshness(board: dict[str, Any]) -> dict[str, Any]:
    """How old the data is, and whether it may still be called live.

    Age is measured from when the provider answered, not from when this request
    was served, so a cache hit does not reset the label to "just now".
    """
    age = board.get("age_seconds")
    age = _int(age) if age is not None else None
    source = str(board.get("source") or "unavailable")
    degraded = source not in {"coingecko"}
    stale = source == "unavailable" or (age is not None and age > STALE_AFTER_SECONDS)
    return {
        "ageSeconds": age,
        "observedAt": board.get("updated_at"),
        "source": source,
        # The only field a client should consult before drawing a "LIVE" dot.
        "live": bool(not stale and not degraded and board.get("markets")),
        "stale": bool(stale),
        "degraded": bool(degraded),
        "warning": board.get("warning") or None,
    }


def global_metrics() -> dict[str, Any]:
    """The market strip: total cap, 24h volume, BTC/ETH dominance, 24h change.

    Read from the briefings overview, which is the only ``/global`` caller in
    the product. Every field is either a provider number or null — the strip
    renders "--" rather than inventing a figure.
    """
    overview = crypto_provider.get_market_overview() or {}
    if not overview:
        return {"available": False, "provider": None, "observedAt": None, "stale": True,
                "totalMarketCap": None, "totalVolume24h": None, "btcDominance": None,
                "ethDominance": None, "marketCapChange24hPct": None, "marketDirection": None}
    stale = crypto_provider.is_stale(overview)
    return {
        # /global is a CoinGecko aggregate; the Coinbase fallback has no
        # equivalent, so under fallback these are null rather than zero.
        "available": overview.get("total_market_cap") is not None,
        "provider": overview.get("provider"),
        "observedAt": overview.get("generated_at"),
        "stale": bool(stale),
        "totalMarketCap": _num(overview.get("total_market_cap")),
        "totalVolume24h": _num(overview.get("total_volume_24h")),
        "btcDominance": _num(overview.get("btc_dominance")),
        "ethDominance": _num(overview.get("eth_dominance")),
        "marketCapChange24hPct": _num(overview.get("market_cap_change_24h_pct")),
        "marketDirection": overview.get("market_direction"),
    }


def _volume_history(symbol: str) -> list[float]:
    """Recorded 24h-volume readings for one symbol, or [] if none are stored.

    ``market_observations`` is written once per ``alert_worker`` cycle from the
    same board this module reads, so this is a local table read, not a provider
    call. When the table is empty the answer is an empty list and the volume
    layer reports UNAVAILABLE — never "normal", which would be an assertion
    about data we do not have.
    """
    try:
        from services import market_observations

        rows = market_observations.get_observations(symbol, limit=48) or []
    except Exception as exc:  # noqa: BLE001 - volume history is optional depth
        LOGGER.info("Volume history unavailable for %s: %s", symbol, exc)
        return []
    values = []
    for row in rows:
        volume = row.get("volume_24h") if isinstance(row, dict) else None
        number = _num(volume)
        if number is not None:
            values.append(number)
    # ``get_observations`` returns oldest first, which is the order the anomaly
    # scan expects: the last element is the current reading and everything
    # before it is the baseline it is unusual (or not) against.
    return values


def _intelligence_context(board: dict[str, Any]) -> dict[str, Any]:
    """Board-wide context computed once, then shared by every row.

    Breadth, the median 24h change and the BTC benchmark are properties of the
    whole board, so computing them per asset would be fifty times the work for
    the same answer — and would let two rows in one response disagree about what
    the market is doing.
    """
    rows = board.get("markets") or []
    context = market_intelligence.board_context(rows, global_metrics())
    context["observedAt"] = board.get("updated_at")
    return context


def _attach_intelligence(assets: list[dict[str, Any]], board: dict[str, Any],
                         context: dict[str, Any] | None = None) -> list[dict[str, Any]]:
    """Row-level verdicts for a whole list, from data already in hand.

    The series each verdict is built on is the un-thinned 7-day hourly sparkline
    that arrived with the board request (``market_data.hourly_series``). No
    branch of this function can make a network call, which is what keeps a
    fifty-row list at one provider call rather than fifty.
    """
    context = context or _intelligence_context(board)
    for asset in assets:
        symbol = str(asset.get("symbol") or "")
        series = market_data.hourly_series(symbol) or asset.get("sparkline") or []
        try:
            asset["intelligence"] = market_intelligence.assess(
                asset, series, context, volume_series=None, holding=None, depth="list",
            )
        except Exception as exc:  # noqa: BLE001 - a verdict must never break the list
            LOGGER.info("Intelligence unavailable for %s: %s", symbol, exc)
            asset["intelligence"] = None
    return assets


def market_rows(category: str = "all", limit: int = 50) -> dict[str, Any]:
    """The asset list for one chip, plus how fresh it is."""
    category = str(category or "all").strip().lower()
    if category not in CATEGORIES:
        category = "all"
    board = _board(category, limit)
    updated_at = board.get("updated_at")
    assets = [a for a in (normalize_asset(row, updated_at) for row in board.get("markets") or []) if a]
    _attach_intelligence(assets, board)
    return {"category": category, "assets": assets, "freshness": _freshness(board)}


def trending() -> dict[str, Any]:
    """CoinGecko's own trending list, priced from the shared board.

    Trending is a provider signal about what people are *searching for*. It is
    not "today's biggest movers" — deriving it from price change would put a
    label on the screen that the provider never asserted. A trending coin that
    is outside the 50-row board keeps a null price rather than being dropped or
    guessed at.
    """
    board = _board("all", 80)
    index = {}
    for row in board.get("markets") or []:
        symbol = str(row.get("symbol") or "").upper()
        if symbol:
            index.setdefault(symbol, row)
    updated_at = board.get("updated_at")
    rows = []
    try:
        coins = crypto_provider.get_trending()
    except Exception as exc:  # noqa: BLE001 - trending is optional colour, never fatal
        LOGGER.info("Market Pulse trending unavailable: %s", exc)
        coins = []
    for coin in coins:
        symbol = str(coin.get("symbol") or "").upper()
        if not symbol:
            continue
        asset = normalize_asset(index.get(symbol), updated_at) or {
            "id": symbol.lower(), "symbol": symbol, "name": coin.get("name") or symbol,
            "rank": _int(coin.get("rank")), "price": None, "change24h": None,
            "marketCap": None, "volume24h": None, "image": "", "sparkline": [],
            "updatedAt": updated_at,
        }
        asset["trending"] = True
        rows.append(asset)
    # A trending coin outside the board has a null price and no series, so its
    # verdict comes back DATA_UNAVAILABLE rather than being derived from the one
    # thing we know about it, which is that people are searching for it.
    _attach_intelligence(rows, board)
    return {
        "category": "trending",
        "assets": rows,
        # Named so nobody later mistakes this for a price-derived list.
        "basis": "coingecko_search_trending",
        "freshness": _freshness(board),
    }


def snapshot(category: str = "all", limit: int = 50) -> dict[str, Any]:
    """One round trip for the whole screen: strip, list, and freshness."""
    started = time.time()
    payload = trending() if category == "trending" else market_rows(category, limit)
    payload["global"] = global_metrics()
    # Regime and rotation come from breadth across the board that was already
    # fetched to draw the list — no second provider call, and no paid breadth
    # feed. They sit alongside the global strip rather than inside it so the
    # existing strip fields stay exactly where every client expects them.
    board = _board("all", 80)
    context = _intelligence_context(board)
    payload["regime"] = context.get("regime")
    payload["rotation"] = context.get("rotation")
    payload["categories"] = list(CATEGORIES)
    payload["ranges"] = list(market_data.HISTORY_RANGES)
    payload["rangeAliases"] = dict(market_data.HISTORY_RANGE_ALIASES)
    payload["elapsedMs"] = int((time.time() - started) * 1000)
    return payload


def asset_history(symbol: str, range_key: str = "24H") -> dict[str, Any]:
    """Real price history. Ranges are cached per (coin, range) by market_data."""
    return market_data.asset_history(symbol, range_key)


def search(query: str, limit: int = 25) -> dict[str, Any]:
    """Search the same board that supplies prices.

    Offering an asset the price engine has never heard of would let a user open
    a detail screen that can only ever read "Unavailable". Symbol is not a
    CoinGecko id, and resolution stays server-side in ``coingecko_client`` and
    the board — there is no second resolver.
    """
    needle = str(query or "").strip().lower()
    board = _board("all", 80)
    updated_at = board.get("updated_at")
    matches = []
    for row in board.get("markets") or []:
        symbol = str(row.get("symbol") or "").lower()
        name = str(row.get("name") or "").lower()
        coin_id = str(row.get("id") or "").lower()
        if not needle or needle in symbol or needle in name or needle in coin_id:
            asset = normalize_asset(row, updated_at)
            if asset:
                matches.append(asset)
    matches.sort(key=lambda a: (a.get("rank") if a.get("rank") is not None else 9999, a["symbol"]))
    results = matches[: max(1, min(int(limit or 25), 80))]
    _attach_intelligence(results, board)
    return {
        "query": query or "",
        "assets": results,
        "freshness": _freshness(board),
    }


def asset_intelligence(symbol: str, holding: dict[str, Any] | None = None) -> dict[str, Any]:
    """The full drill-down for the one asset a user has open.

    The deep layers cost at most one cached history call — ``asset_history`` is
    memoised per (symbol, range) — and reuse the board's hourly series when it
    already covers the window. That is the whole reason the expensive layers
    live behind a tap: a screen shows fifty rows and one open asset, so the work
    that cannot be shared is only ever done once.

    ``holding`` must be real portfolio data supplied by the caller. This
    function never guesses whether someone owns the asset; with no holding it
    returns the non-holder framing and says which one it used.
    """
    symbol = str(symbol or "").upper().strip()
    board = _board("all", 80)
    context = _intelligence_context(board)
    row = next((r for r in board.get("markets") or [] if str(r.get("symbol") or "").upper() == symbol), None)
    asset = normalize_asset(row, board.get("updated_at")) if row else {"symbol": symbol, "price": None}

    series = market_data.hourly_series(symbol)
    price_source = "coingecko board sparkline (7d hourly)"
    if len(series) < market_intelligence.MIN_POINTS_FOR_FULL_DEPTH:
        # The board could not supply a usable series — most often because we are
        # on the Coinbase fallback, which carries no sparkline at all. The 7D
        # chart is already cached for this screen, so reading it here adds no
        # provider call in the common case.
        history = market_data.asset_history(symbol, "7D") or {}
        points = [p.get("price") for p in history.get("points") or []]
        if len(points) > len(series):
            series = [p for p in points if isinstance(p, (int, float))]
            price_source = "coingecko 7d market chart"
    context["priceSource"] = price_source

    payload = market_intelligence.assess(
        asset, series, context,
        volume_series=_volume_history(symbol),
        holding=holding,
        depth="full",
    )
    payload["asset"] = asset
    payload["freshness"] = _freshness(board)
    payload["regime"] = context.get("regime")
    payload["rotation"] = context.get("rotation")
    return payload

File: mobile-native/src/api/marketPulse.ts
/**
 * Market Pulse client — the live crypto command center's only data path.
 *
 * ## The phone never talks to CoinGecko
 *
 * Every read here goes to `/api/pulse/market/*`, which reads the shared market
 * foundation the dashboard board and Pulse Briefings already poll. One provider
 * call per window serves every user, so user count does not scale provider
 * spend. There is deliberately no CoinGecko host, no API key and no coin-id
 * table in this file: a symbol is not a CoinGecko id, and resolving one to the
 * other is server work. A second resolver in native code would drift from the
 * first the day a coin is relisted under a new id.
 *
 * ## Absence is null, all the way to the pixel
 *
 * `Number(null)` is `0`, and `$0.00` beside `0.00%` reads on screen as a fact
 * about the market rather than as a gap in the data. So the server sends null
 * for "we do not know" and this file's whole job on the way in is to keep that
 * null intact for `formatPrice` to render as "--". There is no last-known-price
 * fallback and no percentage computed from two numbers we happen to hold.
 *
 * ## Freshness is a server claim, not a client guess
 *
 * `freshness.live` is the only field a screen may consult before drawing a LIVE
 * dot, and `ageSeconds` is measured from when the provider answered — not from
 * when this request was served, so a cache hit cannot reset "Updated 12s ago"
 * to zero. A client that timed its own fetch would report the age of the
 * request instead of the age of the price.
 *
 * ## Writes are not here
 *
 * Adding to a watchlist and creating an alert go through the existing crypto
 * API (`api/watchlists.ts`, `api/alerts.ts`). There is one watchlist store and
 * one alert engine in this product, and Market Pulse is a reader of both.
 */

import { readJsonCache, writeJsonCache } from "../core/cache";
import {
  AssetIntelligence,
  MarketRegime,
  MarketRotation,
  normalizeIntelligence,
  normalizeRegime,
  normalizeRotation
} from "./marketIntelligence";
import { pulseApi } from "./pulseApi";

const SNAPSHOT_CACHE_PREFIX = "pulsesoc.native.marketpulse.snapshot.";

/**
 * The chips, in display order.
 *
 * Only categories the canonical provider data can actually answer. "trending"
 * is CoinGecko's own search-trending signal rather than a sort of the price
 * board — labelling today's biggest mover "trending" would be a claim the
 * provider never made.
 */
export const MARKET_CATEGORIES = ["all", "gainers", "losers", "trending", "watchlist"] as const;
export type MarketCategory = (typeof MARKET_CATEGORIES)[number];

/**
 * Chart ranges the backend can answer, in the spelling the chart shows.
 *
 * 30D and 90D are aliases the server resolves to its stored 1M/3M keys, so no
 * existing caller, cache entry or saved preference has to change spelling for
 * the chart to read the way a trader expects. ALL is bounded server-side by the
 * provider plan; the client never asks for `max`.
 */
export const PULSE_RANGES = ["1H", "24H", "7D", "30D", "90D", "1Y", "ALL"] as const;
export type PulseRange = (typeof PULSE_RANGES)[number];

/** One asset, in PulseSoc's own shape. No CoinGecko field name survives. */
export type MarketAsset = {
  id: string;
  symbol: string;
  name: string;
  rank: number | null;
  price: number | null;
  change24h: number | null;
  marketCap: number | null;
  volume24h: number | null;
  image: string;
  sparkline: number[];
  updatedAt: string | null;
  /** Per-account, merged server-side inside one authenticated request. */
  watching: boolean;
  favorite: boolean;
  alertCount: number;
  /** Set only on rows that came from the provider's trending list. */
  trending: boolean;
  /**
   * The server's verdict for this row, or null when it could not form one.
   *
   * Null is a real state and is not the same as a cautious verdict: the row's
   * price is still true, the analysis simply had too little history. A card
   * renders nothing rather than a placeholder opinion.
   */
  intelligence: AssetIntelligence | null;
};

export type MarketFreshness = {
  ageSeconds: number | null;
  observedAt: string | null;
  source: string;
  /** The single field a screen may consult before claiming to be live. */
  live: boolean;
  stale: boolean;
  degraded: boolean;
  warning: string;
};

export type GlobalMetrics = {
  available: boolean;
  provider: string;
  observedAt: string | null;
  stale: boolean;
  totalMarketCap: number | null;
  totalVolume24h: number | null;
  btcDominance: number | null;
  ethDominance: number | null;
  marketCapChange24hPct: number | null;
  marketDirection: string;
};

export type MarketSnapshot = {
  ok: boolean;
  category: MarketCategory;
  assets: MarketAsset[];
  freshness: MarketFreshness;
  global: GlobalMetrics;
  /** Present on the trending payload so nobody mistakes it for a price sort. */
  basis: string;
  /**
   * False when the per-account overlay could not be read. The market list is
   * still true without the badges, so the screen stays up and simply does not
   * claim to know what the user is watching.
   */
  personalized: boolean;
  /**
   * Conditions across the whole board, computed from breadth the snapshot
   * already contains — no extra provider call and no paid breadth feed. Null
   * when the board was too small to measure it.
   */
  regime: MarketRegime | null;
  rotation: MarketRotation | null;
};

export type PulseHistoryPoint = { t: number; price: number };

export type PulseHistory = {
  ok: boolean;
  symbol: string;
  range: string;
  points: PulseHistoryPoint[];
  source: string;
  warning: string;
  stale: boolean;
};

type Raw<T> = T extends (infer U)[] ? Raw<U>[] : T extends object ? { [K in keyof T]?: Raw<T[K]> } : T;

/** A number, or null when the server had nothing. `0` is a price; null is not. */
function nullableNumber(value: unknown): number | null {
  if (value === null || value === undefined || value === "") return null;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

function nullableText(value: unknown): string | null {
  if (value === null || value === undefined) return null;
  const text = String(value).trim();
  return text ? text : null;
}

function sparklinePoints(value: unknown): number[] {
  if (!Array.isArray(value)) return [];
  // A gap in the provider's series is dropped rather than plotted. `Number(null)`
  // is `0` and passes a finiteness check, so the obvious spelling of this filter
  // would draw every gap as a crash to zero — a much louder lie than a slightly
  // compressed shape.
  return value.map(nullableNumber).filter((point): point is number => point !== null);
}

export function normalizeMarketAsset(input: Raw<MarketAsset> = {}): MarketAsset {
  const symbol = String(input.symbol || "").toUpperCase();
  return {
    id: String(input.id || symbol.toLowerCase()),
    symbol,
    name: String(input.name || symbol),
    rank: nullableNumber(input.rank),
    price: nullableNumber(input.price),
    change24h: nullableNumber(input.change24h),
    marketCap: nullableNumber(input.marketCap),
    volume24h: nullableNumber(input.volume24h),
    image: String(input.image || ""),
    sparkline: sparklinePoints(input.sparkline),
    updatedAt: nullableText(input.updatedAt),
    watching: Boolean(input.watching),
    favorite: Boolean(input.favorite),
    // A badge count, not a price: 0 is the honest answer for "no alerts".
    alertCount: Math.max(0, Number(input.alertCount) || 0),
    trending: Boolean(input.trending),
    // Owned by `marketIntelligence.ts`, because the same shape arrives from two
    // endpoints and one normalizer for it is one place to keep it honest.
    intelligence: normalizeIntelligence(input.intelligence)
  };
}

export function normalizeFreshness(input: Raw<MarketFreshness> = {}): MarketFreshness {
  return {
    ageSeconds: nullableNumber(input.ageSeconds),
    observedAt: nullableText(input.observedAt),
    source: String(input.source || "unavailable"),
    live: Boolean(input.live),
    // An unreadable payload is treated as stale rather than as fresh. The
    // failure mode of the opposite default is a LIVE dot over old numbers.
    stale: input.stale === undefined ? true : Boolean(input.stale),
    degraded: Boolean(input.degraded),
    warning: String(input.warning || "")
  };
}

export function normalizeGlobalMetrics(input: Raw<GlobalMetrics> = {}): GlobalMetrics {
  return {
    available: Boolean(input.available),
    provider: String(input.provider || ""),
    observedAt: nullableText(input.observedAt),
    stale: input.stale === undefined ? true : Boolean(input.stale),
    totalMarketCap: nullableNumber(input.totalMarketCap),
    totalVolume24h: nullableNumber(input.totalVolume24h),
    btcDominance: nullableNumber(input.btcDominance),
    ethDominance: nullableNumber(input.ethDominance),
    marketCapChange24hPct: nullableNumber(input.marketCapChange24hPct),
    marketDirection: String(input.marketDirection || "unknown")
  };
}

function isCategory(value: unknown): value is MarketCategory {
  return (MARKET_CATEGORIES as readonly string[]).includes(String(value));
}

export function normalizeSnapshot(input: Raw<MarketSnapshot> = {}): MarketSnapshot {
  return {
    ok: Boolean(input.ok),
    category: isCategory(input.category) ? (input.category as MarketCategory) : "all",
    assets: (input.assets || []).map(normalizeMarketAsset).filter((asset) => Boolean(asset.symbol)),
    freshness: normalizeFreshness(input.freshness || {}),
    global: normalizeGlobalMetrics(input.global || {}),
    basis: String(input.basis || ""),
    personalized: Boolean(input.personalized),
    regime: normalizeRegime(input.regime),
    rotation: normalizeRotation(input.rotation)
  };
}

export function normalizeHistory(input: Raw<PulseHistory> = {}): PulseHistory {
  const points = (input.points || [])
    .map((point) => ({ t: nullableNumber(point?.t), price: nullableNumber(point?.price) }))
    .filter((point): point is PulseHistoryPoint => point.t !== null && point.price !== null);
  return {
    ok: Boolean(input.ok) && points.length > 0,
    symbol: String(input.symbol || "").toUpperCase(),
    range: String(input.range || "24H").toUpperCase(),
    points,
    source: String(input.source || "unavailable"),
    warning: String(input.warning || ""),
    stale: Boolean(input.stale)
  };
}

// ---------------------------------------------------------------------------
// Reads
// ---------------------------------------------------------------------------

/**
 * Everything the screen needs on open, in one request.
 *
 * Pull-to-refresh calls exactly this. It does not carry a bypass flag, because
 * the cache, the single-flight and the credit guard it would bypass are the
 * only reasons the provider bill does not scale with how hard users pull.
 */
export async function getMarketSnapshot(category: MarketCategory = "all", limit = 50) {
  const snapshot = normalizeSnapshot(
    await pulseApi<MarketSnapshot>(
      `/api/pulse/market/snapshot?category=${encodeURIComponent(category)}&limit=${encodeURIComponent(String(limit))}`
    )
  );
  // Only a snapshot with rows is worth keeping. Caching an empty outage payload
  // would make a thirty-second provider blip look permanent on next launch.
  if (snapshot.assets.length) {
    await writeJsonCache(`${SNAPSHOT_CACHE_PREFIX}${category}`, snapshot).catch(() => undefined);
  }
  return snapshot;
}

export async function loadCachedMarketSnapshot(category: MarketCategory = "all") {
  return readJsonCache<MarketSnapshot>(`${SNAPSHOT_CACHE_PREFIX}${category}`, normalizeSnapshot);
}

/** The global strip alone — a cheaper foreground refresh than the whole board. */
export async function getGlobalMetrics() {
  const response = await pulseApi<{ ok?: boolean; global?: Raw<GlobalMetrics> }>("/api/pulse/market/global");
  return normalizeGlobalMetrics(response.global || {});
}

/**
 * Search the same board that supplies prices.
 *
 * Every hit is therefore openable. Offering a coin the price engine has never
 * heard of would hand the user a detail screen that can only read "Unavailable".
 */
export async function searchMarketAssets(query: string, limit = 25) {
  const response = await pulseApi<{ ok?: boolean; assets?: Raw<MarketAsset>[] }>(
    `/api/pulse/market/search?q=${encodeURIComponent(query)}&limit=${encodeURIComponent(String(limit))}`
  );
  return (response.assets || []).map(normalizeMarketAsset).filter((asset) => Boolean(asset.symbol));
}

/** Real price history for one asset and range. Never a synthesised series. */
export async function getPulseHistory(symbol: string, range: PulseRange = "24H") {
  return normalizeHistory(
    await pulseApi<PulseHistory>(
      `/api/pulse/market/assets/${encodeURIComponent(symbol.toUpperCase())}/history?range=${encodeURIComponent(range)}`
    )
  );
}

// ---------------------------------------------------------------------------
// Display helpers
//
// The single place "we do not know" becomes "--", so no screen has to remember
// to check for null before formatting.
// ---------------------------------------------------------------------------

export const UNKNOWN = "--";

export function formatBigMoney(value: number | null): string {
  if (value === null) return UNKNOWN;
  const abs = Math.abs(value);
  if (abs >= 1e12) return `$${(value / 1e12).toFixed(2)}T`;
  if (abs >= 1e9) return `$${(value / 1e9).toFixed(2)}B`;
  if (abs >= 1e6) return `$${(value / 1e6).toFixed(2)}M`;
  return `$${value.toLocaleString(undefined, { maximumFractionDigits: 0 })}`;
}

export function formatDominance(value: number | null): string {
  return value === null ? UNKNOWN : `${value.toFixed(1)}%`;
}

/**
 * "Updated 12s ago" — or nothing at all when the age is unknown.
 *
 * A null age means the server could not say when the price was observed, and
 * "Updated just now" would be the app inventing the one number this label
 * exists to report.
 */
export function formatAge(ageSeconds: number | null): string {
  if (ageSeconds === null || ageSeconds < 0) return "";
  if (ageSeconds < 60) return `Updated ${Math.floor(ageSeconds)}s ago`;
  if (ageSeconds < 3600) return `Updated ${Math.floor(ageSeconds / 60)}m ago`;
  if (ageSeconds < 86400) return `Updated ${Math.floor(ageSeconds / 3600)}h ago`;
  return `Updated ${Math.floor(ageSeconds / 86400)}d ago`;
}
